cps/elf.c

```diff
@@ -0,0 +1,406 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <elf.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <string.h>
+#include <fcntl.h>
+
+#include "gatekeeper_cps.h"
+#include "elf.h"
+
+#if __WORDSIZE == 32
+#define PERBIT(x) x##32
+#define ElfPERBIT(x) Elf32_##x
+#else
+#define PERBIT(x) x##64
+#define ElfPERBIT(x) Elf64_##x
+#endif
+
+struct elf_file {
+	/* Path to ELF file. */
+	char *pathname;
+
+	/* Whether the endianness needs to be converted. */
+	int  conv;
+
+	/* File contents. */
+	void *data;
+
+	/* Length of file contents. */
+	unsigned long len;
+};
+
+struct string_table
+{
+	/* Number of strings in the table. */
+	unsigned int cnt;
+
+	/* Maximum number of strings allowed in table. */
+	unsigned int max;
```

Comment:
It's better to use another name instead of ``max``: (1) to more accurately represent the meaning; (2) avoid confusion as in other languages, ``max`` typically is a function name.